--- ./configure~	2002-07-28 06:50:42.000000000 +0300
+++ ./configure	2003-04-13 17:20:53.000000000 +0300
@@ -5366,13 +5366,13 @@
 
         case "$host" in
     i?86-* | k?-*)
-	case "$host" in
-	i386-*) TRY_CFLAGS="$OPT_CFLAGS -mcpu=i386";;
-	i486-*) TRY_CFLAGS="$OPT_CFLAGS -mcpu=i486";;
-	i586-*) TRY_CFLAGS="$OPT_CFLAGS -mcpu=pentium";;
-	i686-*) TRY_CFLAGS="$OPT_CFLAGS -mcpu=pentiumpro";;
-	k6-*)   TRY_CFLAGS="$OPT_CFLAGS -mcpu=k6";;
-	esac
+#	case "$host" in
+#	i386-*) TRY_CFLAGS="$OPT_CFLAGS -mcpu=i386";;
+#	i486-*) TRY_CFLAGS="$OPT_CFLAGS -mcpu=i486";;
+#	i586-*) TRY_CFLAGS="$OPT_CFLAGS -mcpu=pentium";;
+#	i686-*) TRY_CFLAGS="$OPT_CFLAGS -mcpu=pentiumpro";;
+#	k6-*)   TRY_CFLAGS="$OPT_CFLAGS -mcpu=k6";;
+#	esac
 	{ printf "%s\n" "$as_me:${as_lineno-$LINENO}: checking if $CC supports $TRY_CFLAGS $CFLAGS flags" >&5
 printf %s "checking if $CC supports $TRY_CFLAGS $CFLAGS flags... " >&6; }
    SAVE_CFLAGS="$CFLAGS"
